sample-api-server/api_home/usr/lib/cgi-bin/api.py

```diff
@@ -0,0 +1,157 @@
+            currentState["scalingStatus"] = "Failed"
+        else:
+            currentState["actualCapacity"] = currentState["desiredCapacity"]
+            currentState["scalingStatus"] = "Successful"
+    return(currentState)
+
+def patch_state(currentState,bodyData):
+    if "actualCapacity" in bodyData:
+        currentState["actualCapacity"] = bodyData["actualCapacity"]
+    if "desiredCapacity" in bodyData:
+        currentState["desiredCapacity"] = bodyData["desiredCapacity"]
+    if "scalingStatus" in bodyData:
+        currentState["scalingStatus"] = bodyData["scalingStatus"]
+    return(currentState)
+
+def end_routine(currentState):
+    global previousState
+    write_state(currentState)
+    print("Content-type: application/json\n\n")
+    print(json.dumps(currentState))
+    logResponse = ""
+    if previousState["desiredCapacity"] != currentState["desiredCapacity"]:
+        logResponse += "desiredCapacity: " + str(previousState["desiredCapacity"]) + " -> " + str(currentState["desiredCapacity"]) + "  "
+    if previousState["actualCapacity"] != currentState["actualCapacity"]:
+        logResponse += "actualCapacity: " + str(previousState["actualCapacity"]) + " -> " + str(currentState["actualCapacity"]) + "  "
+    if previousState["scalingStatus"] != currentState["scalingStatus"]:
+        logResponse += "scalingStatus: " + previousState["scalingStatus"] + " -> " + currentState["scalingStatus"] 
+    if not logResponse:
+        logResponse += "actualCapacity: " + str(currentState["actualCapacity"]) + " = desiredCapacity: " + str(currentState["desiredCapacity"]) + "  "
+    if bodyData:
+        bodyOut = json.dumps(bodyData)
+    else:
+        bodyOut = ""
+    logData = datetime.now().strftime("%Y%m%d-%H%M%S") + " Request: " + os.environ["REQUEST_METHOD"] + " dimensionId: " + dimensionId +" "  + bodyOut + "\n"
+    logData += datetime.now().strftime("%Y%m%d-%H%M%S") + " Response: dimensionId: " + dimensionId + "  " + logResponse  + "\n"
+    logFile = open(logFileName , "a")
+    logFile.write(logData)
+    logFile.close()
+    if debugOn:
+        debugLog = open(debugLogFile, "a")
+        for param in os.environ.keys():
+            debugLog.write(str(param) + " " + str(os.environ[param]) + "\n")
+        debugLog.write(json.dumps(bodyData)+"\n")
+        debugLog.close()
+    exit() 
+
+### Main script logic
+bodyData = None
+manage_state_file()
+currentState = read_state()
+global previousState
+previousState = dict(currentState)
+## Grab body from PATCH
+if "CONTENT_LENGTH" in os.environ:
+    contentLength = int(os.environ["CONTENT_LENGTH"])
+    bodyData = json.loads(sys.stdin.read(contentLength))
+if bodyData and os.environ["REQUEST_METHOD"] == 'PATCH':
+    if demoMode:
+        ## For desiredCapacity change, auto-set status to pending.
+        if (bodyData["desiredCapacity"]) != (currentState["desiredCapacity"]):
```

Comment:
"desiredCapacity" in bodyData ?

Reply:
The PATCH request body is set to a dict 'bodyData'.  
That request body sometimes includes the desiredCapacity key-value pair.

Reply:
1. In demoMode, would 'bodyData' always contains 'desiredCapacity'? The reason i ask this is because in non-demo mode we do this check at line number 141. Apologies I am not an expert in python and if the current code already handles it somehow.
2. Do we have any unit tests for this code. We are planning to integrate with Travis CI or AWS CodeBuild. I am hopeful that if we start adding unit tests we could introduce changes without breaking anything

Reply:
1) Yes.  Remember the behavior of AWS Application autoscaling remains the same regardless of demo mode.   
The mode changes whether the sample api server automatically rotates through the available states to simulate a successful scaling event.
The only automatic state change made in demo mode is to respond to the initial scaling request with a pending state.  That's what you see in line 141.  
2)  We could build and run the Docker container in a CiCd platform and call it on localhost to test the response.   
curl http://localhost/v1/scalableTargetDimensions/1-23456789  
HTTPS would be much more difficult to set up.

Reply:
I'm sure we can devise tests that fit your needs.  I would want to have a deeper discussion to see how this would integrate with your current test plans.

Reply:
shouldn't unit tests be independent of what continuous build system we might use. I am curious if our plan of adding unit tests be affected by the choice of it - ex. AWS CodeBuild vs Travis CI
Is there some other aspect to it that you want to discuss.